### PR TITLE
t8n output processing

### PR DESCRIPTION
**Pull request by yerke26**
processing t8n output, handling exceptions, setting exit codes

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [ ] No

#### Notes on testing

_Optional. Remove if not applicable._

#### Requires documentation update

- [ ] Yes
- [ ] No

_If yes, link the PR to the docs update or the issue with the details labeled `docs`. Remove if not applicable._

#### Requires explanation in Release Notes

- [ ] Yes
- [ ] No

_If yes, fill in the details here. Remove if not applicable._

## Remarks

_Optional. Remove if not applicable._

### CODE REVIEW

File: tools/Evm/T8n/JsonConverters/AccountStateJsonConverter.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        NumberConversion? previousValue = ForcedNumberConversion.ForcedConversion.Value;
+        try
+        {
+            writer.WriteStartObject();
+
+            ForcedNumberConversion.ForcedConversion.Value = NumberConversion.Hex;
+            writer.WritePropertyName("balance"u8);
+            JsonSerializer.Serialize(writer, value.Balance, options);
+
+            ForcedNumberConversion.ForcedConversion.Value = NumberConversion.Hex;
```
Comment (LukaszRozmej) [inline tools/Evm/T8n/JsonConverters/AccountStateJsonConverter.cs:34]: No need, it is already forced?

File: tools/Evm/T8n/JsonConverters/AccountStateJsonConverter.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+using System.Text.Json.Serialization;
+using Ethereum.Test.Base;
+using Nethermind.Core.Extensions;
+using Nethermind.Int256;
+using Nethermind.Serialization.Json;
+
+namespace Evm.T8n.JsonConverters;
+
+// required to serialize in geth t8n format
+public class AccountStateJsonConverter : JsonConverter<AccountState>
```
Comment (LukaszRozmej) [inline tools/Evm/T8n/JsonConverters/AccountStateJsonConverter.cs:14]: Do we really need it? Wouldn't just setting some attributes on `AccountState` suffice?
I think we should move to serializing numbers same as Geth! @MarekM25 @smartprogrammer93 ?

File: tools/Evm/T8n/JsonConverters/ReceiptJsonConverter.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        writer.WriteStartObject();
+
+        if (receipt.TxType != TxType.Legacy)
+        {
+            writer.WritePropertyName("type");
+            JsonSerializer.Serialize(writer, receipt.TxType, options);
+        }
+        writer.WritePropertyName("root");
+        ByteArrayConverter.Convert(writer, receipt.PostTransactionState is not null ? receipt.PostTransactionState.Bytes : Bytes.ZeroByte.ToArray());
```
Comment (LukaszRozmej) [inline tools/Evm/T8n/JsonConverters/ReceiptJsonConverter.cs:36]: Don't allocate new array

File: tools/Evm/T8n/JsonConverters/ReceiptJsonConverter.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Extensions;
+using Nethermind.Int256;
+using Nethermind.Serialization.Json;
+
+namespace Evm.T8n.JsonConverters;
+
+// required to serialize in geth t8n format
+public class ReceiptJsonConverter : JsonConverter<TxReceipt>
```
Comment (LukaszRozmej) [inline tools/Evm/T8n/JsonConverters/ReceiptJsonConverter.cs:17]: Can we achieve Geth format without converter? What about `ReceiptForRpc`?

File: tools/Evm/T8n/T8nCommand.cs [hunk 5]
```diff
@@ -15,9 +29,67 @@ public static void Configure(ref CliRootCommand rootCmd)
...
+
+                if (arguments.OutputResult == Stdout) t8nOutput.Result = t8nExecutionResult.PostState;
+                else WriteToFile(arguments.OutputResult, arguments.OutputBaseDir, t8nExecutionResult.PostState);
+
+                if (arguments.OutputBody == Stdout) t8nOutput.Body = t8nExecutionResult.TransactionsRlp;
+                else if (arguments.OutputBody is not null)
+                {
+                    WriteToFile(arguments.OutputBody, arguments.OutputBaseDir, t8nExecutionResult.TransactionsRlp);
+                }
+
```
Comment (LukaszRozmej) [inline tools/Evm/T8n/T8nCommand.cs:48]: Can you do it in one `if-else`?

File: tools/Evm/T8n/T8nCommand.cs [hunk 6]
```diff
@@ -15,9 +29,67 @@ public static void Configure(ref CliRootCommand rootCmd)
...
+                if (t8nOutput.Body is not null || t8nOutput.Alloc is not null || t8nOutput.Result is not null)
```
Comment (LukaszRozmej) [inline tools/Evm/T8n/T8nCommand.cs:49]: Refactor to `!t8nOutput.IsEmpty`

File: tools/Evm/T8n/T8nCommand.cs [hunk 7]
```diff
@@ -15,9 +29,67 @@ public static void Configure(ref CliRootCommand rootCmd)
...
+                t8nOutput = new T8nOutput(e.Message, T8nErrorCodes.ErrorIO);
+            }
+            catch (JsonException e)
+            {
+                t8nOutput = new T8nOutput(e.Message, T8nErrorCodes.ErrorJson);
+            }
+            catch (Exception e)
+            {
+                t8nOutput = new T8nOutput(e.Message, T8nErrorCodes.ErrorEvm);
+            }
```
Comment (LukaszRozmej) [inline tools/Evm/T8n/T8nCommand.cs:69]: You probably want to log full exception somewhere for debugging purposes. Call stack will be useful.

File: tools/Evm/T8n/T8nCommand.cs [hunk 8]
```diff
@@ -15,9 +29,67 @@ public static void Configure(ref CliRootCommand rootCmd)
...
         rootCmd.Add(cmd);
     }
+
+    private static void WriteToFile(string filename, string? basedir, object outputObject)
+    {
+        if (basedir is not null)
+        {
+            basedir = basedir.TrimEnd('/');
+            basedir += '/';
+        }
```
Comment (LukaszRozmej) [inline tools/Evm/T8n/T8nCommand.cs:89]: why needed?

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]

File: tools/Evm/T8n/T8nCommand.cs (hunk 8 above)
Comment (yerke26) [reply tools/Evm/T8n/T8nCommand.cs:89]: basedir can be with and without slash from user

File: tools/Evm/T8n/JsonConverters/AccountStateJsonConverter.cs (hunk 2 above)
Comment (MarekM25) [reply tools/Evm/T8n/JsonConverters/AccountStateJsonConverter.cs:14]: I agree

File: tools/Evm/T8n/T8nCommand.cs (hunk 8 above)
Comment (LukaszRozmej) [reply tools/Evm/T8n/T8nCommand.cs:89]: that shouldn't matter to the rest of the code though, ah I see the problem:

instead `FileInfo fileInfo = new(basedir + filename);` do `FileInfo fileInfo = new(Path.Combine(basedir, filename));` and you don't need to worry about separators

File: tools/Evm/T8n/T8nCommand.cs (hunk 5 above)
Comment (yerke26) [reply tools/Evm/T8n/T8nCommand.cs]: used GetOrWriteToFile general function

File: tools/Evm/T8n/JsonConverters/ReceiptJsonConverter.cs (hunk 4 above)
Comment (LukaszRozmej) [reply tools/Evm/T8n/JsonConverters/ReceiptJsonConverter.cs]: Can you explain in detail what are the differences? Can we just inherit from it and change something?

Comment (yerke26) [reply tools/Evm/T8n/JsonConverters/ReceiptJsonConverter.cs]:
```
type Receipt struct {
	// Consensus fields: These fields are defined by the Yellow Paper
	Type              uint8  `json:"type,omitempty"`
	PostState         []byte `json:"root"`
	Status            uint64 `json:"status"`
	CumulativeGasUsed uint64 `json:"cumulativeGasUsed" gencodec:"required"`
	Bloom             Bloom  `json:"logsBloom"         gencodec:"required"`
	Logs              []*Log `json:"logs"              gencodec:"required"`

	// Implementation fields: These fields are added by geth when processing a transaction.
	TxHash            common.Hash    `json:"transactionHash" gencodec:"required"`
	ContractAddress   common.Address `json:"contractAddress"`
	GasUsed           uint64         `json:"gasUsed" gencodec:"required"`
	EffectiveGasPrice *big.Int       `json:"effectiveGasPrice"` // required, but tag omitted for backwards compatibility
	BlobGasUsed       uint64         `json:"blobGasUsed,omitempty"`
	BlobGasPrice      *big.Int       `json:"blobGasPrice,omitempty"`

	// Inclusion information: These fields provide information about the inclusion of the
	// transaction corresponding to this receipt.
	BlockHash        common.Hash `json:"blockHash,omitempty"`
	BlockNumber      *big.Int    `json:"blockNumber,omitempty"`
	TransactionIndex uint        `json:"transactionIndex"`
}
```
here omitempty means to omit a field from being included in the JSON output if the field has an empty value

nethermind receipts:
```
[
  {
    "transactionHash": "0xa98a24882ea90916c6a86da650fbc6b14238e46f0af04a131ce92be897507476",
    "transactionIndex": "0x0",
    "blockNumber": "0x0",
    "cumulativeGasUsed": "0x84d0",
    "gasUsed": "0x84d0",
    "from": "0xd02d72e067e77158444ef2020ff2d325f929b363",
    "to": "0x1111111111111111111111111111111111111111",
    "contractAddress": null,
    "logs": [],
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "status": "0x0",
    "error": "BadInstruction",
    "type": "0x2"
  },
  {
    "transactionHash": "0x36bad80acce7040c45fd32764b5c2b2d2e6f778669fb41791f73f546d56e739a",
    "transactionIndex": "0x1",
    "blockNumber": "0x0",
    "cumulativeGasUsed": "0x109a0",
    "gasUsed": "0x84d0",
    "from": "0xd02d72e067e77158444ef2020ff2d325f929b363",
    "to": "0x1111111111111111111111111111111111111111",
    "contractAddress": null,
    "logs": [],
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "status": "0x0",
    "error": "BadInstruction",
    "type": "0x2"
  }
]
```
geth receipts:
```
[
  {
    "type": "0x2",
    "root": "0x",
    "status": "0x0",
    "cumulativeGasUsed": "0x84d0",
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "logs": null,
    "transactionHash": "0xa98a24882ea90916c6a86da650fbc6b14238e46f0af04a131ce92be897507476",
    "contractAddress": "0x0000000000000000000000000000000000000000",
    "gasUsed": "0x84d0",
    "effectiveGasPrice": null,
    "blockHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "transactionIndex": "0x0"
  },
  {
    "type": "0x2",
    "root": "0x",
    "status": "0x0",
    "cumulativeGasUsed": "0x109a0",
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "logs": null,
    "transactionHash": "0x36bad80acce7040c45fd32764b5c2b2d2e6f778669fb41791f73f546d56e739a",
    "contractAddress": "0x0000000000000000000000000000000000000000",
    "gasUsed": "0x84d0",
    "effectiveGasPrice": null,
    "blockHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "transactionIndex": "0x1"
  }
]
```

Comment (LukaszRozmej) [reply tools/Evm/T8n/JsonConverters/ReceiptJsonConverter.cs]: Looks close enough to reuse it with some additional instrumentation? Maybe you can tackle it in the 2nd PR.

**Review by LukaszRozmej — APPROVED** [review]
I think we can remove those json converters in subsequent PR.

**Pull request merged**